protocol/net.go

```diff
@@ -0,0 +1,111 @@
+package protocol
+
+import (
+	"bufio"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-msgio"
+	"github.com/libp2p/go-msgio/protoio"
+	pb "github.com/libp2p/go-smart-record/protocol/pb"
+)
+
+// Idle time before the stream is closed
+var streamIdleTimeout = 1 * time.Minute
+
+// Timeout to wait for a response after a request is sent
+var readMessageTimeout = 10 * time.Second
+
+// The Protobuf writer performs multiple small writes when writing a message.
+// We need to buffer those writes, to make sure that we're not sending a new
+// packet for every single write.
+type bufferedDelimitedWriter struct {
+	*bufio.Writer
+	protoio.WriteCloser
+}
+
+var writerPool = sync.Pool{
+	New: func() interface{} {
+		w := bufio.NewWriter(nil)
+		return &bufferedDelimitedWriter{
+			Writer:      w,
+			WriteCloser: protoio.NewDelimitedWriter(w),
+		}
+	},
+}
+
+func writeMsg(w io.Writer, mes *pb.Message) error {
+	bw := writerPool.Get().(*bufferedDelimitedWriter)
+	bw.Reset(w)
+	err := bw.WriteMsg(mes)
+	if err == nil {
+		err = bw.Flush()
+	}
+	bw.Reset(nil)
+	writerPool.Put(bw)
+	return err
+}
+
+// handleNewStream implements the network.StreamHandler
+func (e *smartRecordServer) handleNewStream(s network.Stream) {
+	if e.handleNewMessage(s) {
+		// If we exited without error, close gracefully.
+		_ = s.Close()
+	} else {
+		// otherwise, send an error.
+		_ = s.Reset()
+	}
+}
+
+// Returns true on orderly completion of writes (so we can Close the stream conveniently).
+func (e *smartRecordServer) handleNewMessage(s network.Stream) bool {
```

Comment:
maybe call it handleAllMessages (plural), because it looks like this is a polling function which continuously reads messages and handles them, if I am reading it correctly. "handleNewMessage" makes it look like its a handler that is called once per message.